extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java

```diff
@@ -0,0 +1,2040 @@
+      MSQSpec querySpec,
+      RowSignature querySignature,
+      ClusterBy queryClusterBy,
+      ColumnMappings columnMappings
+  )
+  {
+    final DataSourceMSQDestination destination = (DataSourceMSQDestination) querySpec.getDestination();
+    final boolean isRollupQ = isRollupQuery(querySpec.getQuery());
+
+    final Pair<List<DimensionSchema>, List<AggregatorFactory>> dimensionsAndAggregators =
+        makeDimensionsAndAggregatorsForIngestion(
+            querySignature,
+            queryClusterBy,
+            destination.getSegmentSortOrder(),
+            columnMappings,
+            isRollupQ,
+            querySpec.getQuery()
+        );
+
+    return new DataSchema(
+        destination.getDataSource(),
+        new TimestampSpec(ColumnHolder.TIME_COLUMN_NAME, "millis", null),
+        new DimensionsSpec(dimensionsAndAggregators.lhs),
+        dimensionsAndAggregators.rhs.toArray(new AggregatorFactory[0]),
+        makeGranularitySpecForIngestion(isRollupQ, querySpec.getQuery()),
+        new TransformSpec(null, Collections.emptyList())
+    );
+  }
+
+  private static GranularitySpec makeGranularitySpecForIngestion(boolean isRollupQ, Query<?> query)
+  {
+    if (!isRollupQ) {
+      return new ArbitraryGranularitySpec(Granularities.NONE, false, Intervals.ONLY_ETERNITY);
+    } else {
+      final String queryGranularity = query.getContextValue(GroupByQuery.CTX_TIMESTAMP_RESULT_FIELD_GRANULARITY, "");
+
+      if (checkIfTimeColumnsAreEqual((GroupByQuery) query) && !queryGranularity.isEmpty()) {
+        return new ArbitraryGranularitySpec(
+            Granularity.fromString(queryGranularity),
+            true,
+            Intervals.ONLY_ETERNITY
+        );
+      }
+      return new ArbitraryGranularitySpec(Granularities.NONE, true, Intervals.ONLY_ETERNITY);
+    }
+  }
+
+
+  /**
+   * Checks if the time columns present in the groupByQuery context are same. One is set by
+   * {@link DruidQuery#toGroupByQuery()} and the other is set by
+   * {@link MSQTaskQueryMaker#runQuery(DruidQuery)}
+   *
+   * @return true if both groupByQuery context values are present and equal else returns false.
+   */
+  private static boolean checkIfTimeColumnsAreEqual(GroupByQuery groupByQuery)
+  {
+    final String msqTimeColumn = groupByQuery.getContextValue(QueryKitUtils.CTX_TIME_COLUMN_NAME, "");
+    if (msqTimeColumn.isEmpty()) {
+      return false;
+    }
+    return msqTimeColumn.equals(groupByQuery.getContextValue(GroupByQuery.CTX_TIMESTAMP_RESULT_FIELD));
+  }
+
+  /**
+   * Checks if segments generated by the insert query can be rolled up futher.
+   *
+   * @param query
+   *
+   * @return
+   */
+  private static boolean isRollupQuery(Query<?> query)
+  {
+    return MultiStageQueryContext.isFinalizeAggregations(query.getQueryContext()) == false
+           && query.getContextBoolean(GroupByQueryConfig.CTX_KEY_ENABLE_MULTI_VALUE_UNNESTING, true) == false
+           && query instanceof GroupByQuery;
```

Comment:
Nit:

```java
return query instanceof GroupByQuery
           && !MultiStageQueryContext.isFinalizeAggregations(query.getQueryContext())
           && 
 !query.getContextBoolean(GroupByQueryConfig.CTX_KEY_ENABLE_MULTI_VALUE_UNNESTING, true);
```

I read this as, "rollup uses a group by query. That query must not finalize aggregations or unnest multi-values." Is unclear to this newbie what it means to finalize aggregations and then unnest; does the SQL syntax support this subtlety?

Reply:
Rewrote it in that way.

Also added this javadoc that should hopefully make the finalization and unnesting requirements more clear:

```
  /**
   * Whether a native query represents an ingestion with rollup.
   *
   * Checks for three things:
   * 
   * - The query must be a {@link GroupByQuery}, because rollup requires columns to be split into dimensions and
   * aggregations.
   * - The query must not finalize aggregations, because rollup requires inserting the intermediate type of
   * complex aggregations, not the finalized type. (So further rollup is possible.)
   * - The query must explicitly disable {@link GroupByQueryConfig#CTX_KEY_ENABLE_MULTI_VALUE_UNNESTING}, because
   * groupBy on multi-value dimensions implicitly unnests, which is not desired behavior for rollup at ingestion time.
   */
```